<a href="https://colab.research.google.com/github/ngving22/NFT-Transaction-Analytics/blob/main/NFT_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Importing original CSV to make the Dataframe**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

#preparing for regression
# import seaborn as sns
# from sklearn import preprocessing, svm
# from sklearn.model_selection import train_test_split
# from sklearn.linear_model import LinearRegression

df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/rarible_exchanges_clean.csv")

df = df.astype( {"assetAmount": float,'assetAddress': str} )

# royalties = df.loc[(df['assetType']=='ETH') & (df['transferType'] == 'ROYALTY')]['assetAmount'].sum()
# payouts = df.loc[(df['assetType']=='ETH') & (df['transferType'] == 'PAYOUT')]['assetAmount'].sum()

# print( f"Number of payouts = {len(df.loc[df['transferType'] == 'PAYOUT'])}" )
# print( f"Number of royalty payments = {len(df.loc[df['transferType'] == 'ROYALTY'])}" )

# print( f"Number of ETH payouts = {len(df.loc[(df['transferType'] == 'PAYOUT')&(df['assetType']=='ETH')])}" )
# print( f"Number of non-ETH payouts = {len(df.loc[(df['transferType'] == 'PAYOUT')&(df['assetType']!='ETH')])}" )
# print( f"Number of ETH royalty payments = {len(df.loc[(df['transferType'] == 'ROYALTY')&(df['assetType']=='ETH')])}" )
# print( f"Number of non-ETH royalty payments = {len(df.loc[(df['transferType'] == 'ROYALTY')&(df['assetType'] != 'ETH')])}" )

# print( f"ETH payouts = {payouts}" )
# print( f"ETH royalties = {royalties}" )
# print( f"Percent royalties = {royalties/payouts}" )

# ##############################

# df2 = pd.DataFrame( { 'assetAddress': df.assetAddress.unique() } )
# totals = df.groupby(['assetAddress','transferType','assetType']).agg({"assetAmount":"sum"})

# payouts = df.loc[(df['assetType']=='ETH') & (df['transferType'] == 'PAYOUT')][['to','assetAmount']].groupby('to').agg({'assetAmount':'sum'}).rename(columns={'assetAmount':'PAYOUT'},inplace=False).sort_values(by='PAYOUT',ascending=False)
# royalties = df.loc[(df['assetType']=='ETH') & (df['transferType'] == 'ROYALTY')][['to','assetAmount']].groupby('to').agg({'assetAmount':'sum'}).rename(columns={'assetAmount':'ROYALTY'},inplace=False).sort_values(by='ROYALTY',ascending=False)
# numpayments = df.loc[(df['assetType']=='ETH') & (df['transferType'] == 'PAYMENT')][['to','assetAmount']].groupby('to').agg({'assetAmount':'count'}).rename(columns={'assetAmount':'PAYOUT'},inplace=False).sort_values(by='PAYOUT',ascending=False)
# numroyalties = df.loc[(df['assetType']=='ETH') & (df['transferType'] == 'ROYALTY')][['to','assetAmount']].groupby('to').agg({'assetAmount':'count'}).rename(columns={'assetAmount':'ROYALTY'},inplace=False).sort_values(by='ROYALTY',ascending=False)

# # payouts = df.loc[([df['transferType']=='PAYOUT'])&(df['assetType']=='ETH')]#.groupby('address').agg({'assetAmount':'sum'})[['address','assetAmount']]


# hist = royalties.hist(bins=np.exp( np.multiply( np.log(10), np.linspace(0,20,40) ) ) )
# plt.savefig("total_royalties.png")
# print( royalties.head() )
# print( royalties.tail() )

# hist = numroyalties.hist(bins=20)
# plt.savefig("count_royalties.png")

# print( numroyalties.head() )
# #df2 = df2.merge( totals.loc[totals.asset

In [ ]:
df.shape[0]

488556

In [ ]:
df.tail()

block_number  log_index            timestamp  \
488551      14767357        428  2022-05-13T12:24:21   
488552      14767357        429  2022-05-13T12:24:21   
488553      14767357        430  2022-05-13T12:24:21   
488554      14767357        431  2022-05-13T12:24:21   
488555      14767357        433  2022-05-13T12:24:21   

                                                to  \
488551  0xF1e1c701B49B1dC2405A9E8Ef40f9f88802B80Fa   
488552  0x1cf0dF2A5A20Cd61d68d4489eEBbf85b8d39e18a   
488553  0x1cf0dF2A5A20Cd61d68d4489eEBbf85b8d39e18a   
488554  0xF1e1c701B49B1dC2405A9E8Ef40f9f88802B80Fa   
488555  0x95317E0Fb4bd69f83d9Fb21972f6ED44833D03c2   

                                              from transferDirection  \
488551  0x95317E0Fb4bd69f83d9Fb21972f6ED44833D03c2          TO_MAKER   
488552  0x95317E0Fb4bd69f83d9Fb21972f6ED44833D03c2          TO_MAKER   
488553  0x95317E0Fb4bd69f83d9Fb21972f6ED44833D03c2          TO_MAKER   
488554  0x95317E0Fb4bd69f83d9Fb21972f6ED44833D03c2          TO_MAKER   
488555  0xF1e1c701B49B1dC2405A9E8Ef40f9f88802B80Fa          TO_TAKER   

       transferType assetType  \
488551      ROYALTY       ETH   
488552       ORIGIN       ETH   
488553       ORIGIN       ETH   
488554       PAYOUT       ETH   
488555       PAYOUT   ERC1155   

                                             assetAddress   assetAmount  \
488551                                                  0  1.100000e+15   
488552                                                  0  2.500000e+14   
488553                                                  0  2.500000e+14   
488554                                                  0  8.650000e+15   
488555  0x7e65a2bc65d5bd26281d5dcbae4c40efe9cce1990000...  1.000000e+00   

       assetSymbol  
488551         NaN  
488552         NaN  
488553         NaN  
488554         NaN  
488555         NaN

**Studying the reOccurrence in "to-address" column**

In [ ]:
df.to.value_counts().head(50)

0x1cf0dF2A5A20Cd61d68d4489eEBbf85b8d39e18a    197445
0x00Ae58bFbF16260832309E1e71D84Fc1B78BD6F5      5039
0x380f997B11b062B4eD20D341DA2730C91492e1f0      4662
0x488B22E75E538473c16bc3A2e950C75C3b0F6700      2575
0xfb8c66c5E4B104c79A1383312a1B52D29b7e8B0b      2428
0x8869e7b48E33c5F1FfFB0F15f6084C7B438D6371      2403
0xc2d131a52a7Bc75F6558588D010d649E02710115      2146
0x279B2309440DC5807eF5e26a0799f2d2CddA94A7      1877
0xD387A6E4e84a6C86bd90C158C6028A58CC8Ac459      1843
0x68dC52058cf885A8782De8d35D0538547F4090C8      1602
0xA59271F4ab69454DbdcE068025F2d1227124E721      1544
0x8Bc9731E393A3112Fe350B6475d7004a2E3545E3      1534
0x981268bF660454e24DBEa9020D57C2504a538C57      1360
0x2218e854019D3Ac7989333470E8CF7BDA8b930F3      1164
0x95Bd75D85C9E9F205d3d6edf3496F8b6864090ab      1149
0x5635828d26B743c62cDCc5081136530d8475b313      1048
0x92458D6BAbE63C464fC5369d56B0dAb864FEabe8      1015
0x21ff1ac88a4A7c07C7573132f976D05B259632EE       996
0x157d90D0e33e90eD1a50e8C8566152d185e2f0b2    

In [ ]:
df["from"].value_counts()

0xC16157e00b1bFf1522C6F01246B4Fb621dA048d0    3840
0x00Ae58bFbF16260832309E1e71D84Fc1B78BD6F5    2262
0xc2d131a52a7Bc75F6558588D010d649E02710115    2146
0x380f997B11b062B4eD20D341DA2730C91492e1f0    2100
0xfA5E2931e3A48209B5D0d5D8F857f97f1818A87B    1232
                                              ... 
0x323fe7fCd33e1A13F89b3913A147Ca6b4bFd07e7       1
0x20859DD4F6Fd38dD64a0EFE77bdb3b35010c3405       1
0x61a035FC087Aaa651161FF1779d305436Fcde1C9       1
0xF3698502A62108aA5D6201cD82256DDb46dbdb85       1
0x093EAb80B444af1FAf373906A3A77cF5BF5A66B7       1
Name: from, Length: 58506, dtype: int64

In [ ]:
df['to'].value_counts().loc[lambda x: x>10000]

0x1cf0dF2A5A20Cd61d68d4489eEBbf85b8d39e18a    197445
Name: to, dtype: int64

In [ ]:
value_counts_to = df['to'].value_counts()

df_to = pd.DataFrame(value_counts_to)
df_to = df_to.reset_index()
df_to.columns = ['address', 'Occurrence']
df_to

address  Occurrence
0      0x1cf0dF2A5A20Cd61d68d4489eEBbf85b8d39e18a      197445
1      0x00Ae58bFbF16260832309E1e71D84Fc1B78BD6F5        5039
2      0x380f997B11b062B4eD20D341DA2730C91492e1f0        4662
3      0x488B22E75E538473c16bc3A2e950C75C3b0F6700        2575
4      0xfb8c66c5E4B104c79A1383312a1B52D29b7e8B0b        2428
...                                           ...         ...
59892  0xA3d4e52b7dDd59373e046343c1aC6EA23490f28f           1
59893  0x503e5A0845cF6e5dF05ee873cEE4BF30D8F9b386           1
59894  0xF9175FBF88Ab9AD0b4E85EF21d3d16759BE82294           1
59895  0xCB8E90F809Ea00AEFC06e34F6A1fE4A3884193Ee           1
59896  0x093EAb80B444af1FAf373906A3A77cF5BF5A66B7           1

[59897 rows x 2 columns]

**At this point, we have 59897 unique to address out of the 488556 transaction**

In [ ]:
df_balance = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/balance_info.csv")

In [ ]:
df_balance

address                   ETH  \
0      0x9E3B4c10DCC866646df0B5BcA2B26221f2B7CC0a    128577399591950122   
1      0x2EEbe34227a023035db26421Af47a2D48DAf6Ca1   1065005506283480341   
2      0x0841f5ADe08074394DFAe3Ca34CF5b5cA527fF7B       564622228958453   
3      0xA372Bf5e08A054FEb5f489f633895DdBf5a8a36A      1531538344808443   
4      0x85EbaB1697e0ce803290A060B15D07Ba058C2bb4     39954074006511444   
...                                           ...                   ...   
59894  0x6DDBF0B85c27167944403f5b2549C464050b30cB     19651954329673381   
59895  0x54E5Cd2F65fE308a8ACbfe37b219478E8cf49C7c       262436595289535   
59896  0xE113E7B1E8Ecd07973EA40978aED520241d17f27  12577825362550968310   
59897  0x8f1D235D4ac923A40F7f9C31DE73822bE8959A05     49732927547379873   
59898  0xe5812192E788b5DA3d762181f90cBd0C8eeEc3B5      3643731919996842   

                  WETH                   RARI   USDC                    DAI  \
0      187500000000000                      0      0                      0   
1                    0                      0      0                      0   
2                    0     441345418396932600      0                      0   
3                    0                      0      0                      0   
4                    0                      0      0                      0   
...                ...                    ...    ...                    ...   
59894                0                      0      0                      0   
59895                0                      0      0                      0   
59896                0  150000000000000000000      0  150000000000000000000   
59897                0                      0      0                      0   
59898                0                      0  13224                      0   

      ASH LUX  FIRST  ARKE ARTx KLTR TRSH                     TATR ROT  
0       0   0      0     0    0    0    0  10000000000000000000000   0  
1       0   0      0     0    0    0    0                        0   0  
2       0   0      0     0    0    0    0                        0   0  
3       0   0      0     0    0    0    0                        0   0  
4       0   0      0     0    0    0    0                        0   0  
...    ..  ..    ...   ...  ...  ...  ...                      ...  ..  
59894   0   0      0     0    0    0    0                        0   0  
59895   0   0      0     0    0    0    0                        0   0  
59896   0   0      0     0    0    0    0                        0   0  
59897   0   0      0     0    0    0    0                        0   0  
59898   0   0      0     0    0    0    0                        0   0  

[59899 rows x 15 columns]

In [ ]:
df_to_balance = pd.merge(df_to, df_balance, how="outer")

In [ ]:
df_to_balance = df_to_balance.apply(pd.to_numeric, errors = 'coerce')

In [ ]:
df3 = pd.DataFrame(df_to_balance.corr())

In [ ]:
df3

address  Occurrence       ETH      WETH      RARI      USDC  \
address         NaN         NaN       NaN       NaN       NaN       NaN   
Occurrence      NaN    1.000000  0.512811  0.294888  0.500244  0.427234   
ETH             NaN    0.512811  1.000000  0.215319  0.267617  0.262496   
WETH            NaN    0.294888  0.215319  1.000000  0.155985  0.160732   
RARI            NaN    0.500244  0.267617  0.155985  1.000000  0.214423   
USDC            NaN    0.427234  0.262496  0.160732  0.214423  1.000000   
DAI             NaN    0.185136  0.138892  0.074675  0.097258  0.093540   
ASH             NaN    0.000286  0.043919  0.004933  0.269089  0.000092   
LUX             NaN    0.000671  0.000120  0.000270  0.000300  0.000193   
FIRST           NaN    0.002864  0.001643  0.000679  0.001315  0.001009   
ARKE            NaN    0.002155  0.000678  0.003145  0.005922  0.000197   
ARTx            NaN    0.000335 -0.000004 -0.000045 -0.000030  0.000082   
KLTR            NaN    0.002710  0.001187  0.000626  0.003555  0.001000   
TRSH            NaN    0.001215  0.001638  0.000197  0.007679  0.000344   
TATR            NaN    0.000036 -0.000005 -0.000107 -0.000032 -0.000078   
ROT             NaN   -0.000055  0.000814  0.002678 -0.000064 -0.000122   

                 DAI       ASH       LUX     FIRST      ARKE      ARTx  \
address          NaN       NaN       NaN       NaN       NaN       NaN   
Occurrence  0.185136  0.000286  0.000671  0.002864  0.002155  0.000335   
ETH         0.138892  0.043919  0.000120  0.001643  0.000678 -0.000004   
WETH        0.074675  0.004933  0.000270  0.000679  0.003145 -0.000045   
RARI        0.097258  0.269089  0.000300  0.001315  0.005922 -0.000030   
USDC        0.093540  0.000092  0.000193  0.001009  0.000197  0.000082   
DAI         1.000000 -0.000068  0.000035  0.000335  0.000254 -0.000064   
ASH        -0.000068  1.000000 -0.000029  0.000022 -0.000028 -0.000028   
LUX         0.000035 -0.000029  1.000000 -0.000029 -0.000025 -0.000017   
FIRST       0.000335  0.000022 -0.000029  1.000000  0.000003 -0.000036   
ARKE        0.000254 -0.000028 -0.000025  0.000003  1.000000 -0.000025   
ARTx       -0.000064 -0.000028 -0.000017 -0.000036 -0.000025  1.000000   
KLTR        0.000345 -0.000048 -0.000032 -0.000062 -0.000045 -0.000032   
TRSH        0.001194  0.000067 -0.000021 -0.000034 -0.000027 -0.000021   
TATR       -0.000081 -0.000029 -0.000018 -0.000035  0.000013 -0.000017   
ROT        -0.000138  0.000246 -0.000029 -0.000061 -0.000041 -0.000028   

                KLTR      TRSH      TATR       ROT  
address          NaN       NaN       NaN       NaN  
Occurrence  0.002710  0.001215  0.000036 -0.000055  
ETH         0.001187  0.001638 -0.000005  0.000814  
WETH        0.000626  0.000197 -0.000107  0.002678  
RARI        0.003555  0.007679 -0.000032 -0.000064  
USDC        0.001000  0.000344 -0.000078 -0.000122  
DAI         0.000345  0.001194 -0.000081 -0.000138  
ASH        -0.000048  0.000067 -0.000029  0.000246  
LUX        -0.000032 -0.000021 -0.000018 -0.000029  
FIRST      -0.000062 -0.000034 -0.000035 -0.000061  
ARKE       -0.000045 -0.000027  0.000013 -0.000041  
ARTx       -0.000032 -0.000021 -0.000017 -0.000028  
KLTR        1.000000  0.103482  0.001028 -0.000025  
TRSH        0.103482  1.000000  0.000736 -0.000011  
TATR        0.001028  0.000736  1.000000 -0.000027  
ROT        -0.000025 -0.000011 -0.000027  1.000000

In [ ]:
df3.to_csv('currency_occurrence_correlation.csv')

In [ ]:
df_to_balance['ETH'].astype(np.float64)
df_to_balance['Occurrence'].astype(np.float64)

0        197445.0
1          5039.0
2          4662.0
3          2575.0
4          2428.0
           ...   
59894         1.0
59895         1.0
59896         1.0
59897         NaN
59898         NaN
Name: Occurrence, Length: 59899, dtype: float64

In [ ]:
column_1 = df_to_balance['ETH'].astype(np.float64)
column_2 = df_to_balance['Occurrence'].astype(np.float64)
ETH_correlation = column_1.corr(column_2)
print(ETH_correlation)

0.5128109552274145


In [ ]:
column_1 = df_to_balance['WETH'].astype(np.float64)
column_2 = df_to_balance['Occurrence'].astype(np.float64)
WETH_correlation = column_1.corr(column_2)
print(WETH_correlation)

0.294888124382395


In [ ]:
column_1 = df_to_balance['RARI'].astype(np.float64)
column_2 = df_to_balance['Occurrence'].astype(np.float64)
RARI_correlation = column_1.corr(column_2)
print(RARI_correlation)

0.5002441695090887


In [ ]:
column_1 = df_to_balance['USDC'].astype(np.float64)
column_2 = df_to_balance['Occurrence'].astype(np.float64)
USDC_correlation = column_1.corr(column_2)
print(USDC_correlation)

0.42723390454351423


In [ ]:
df_to_balance['ETH_base10'] = np.log10(df_to_balance['ETH'].astype(np.float64))

/usr/local/lib/python3.7/dist-packages/pandas/core/arraylike.py:364: RuntimeWarning: divide by zero encountered in log10
  result = getattr(ufunc, method)(*inputs, **kwargs)


In [ ]:
df_to_balance['Occurrence_base2'] = np.log2(df_to_balance['Occurrence'].astype(np.float64))

In [ ]:
df_to_balance

address  Occurrence           ETH          WETH          RARI  \
0          NaN    197445.0  3.679210e+21  6.645540e+20  1.892546e+22   
1          NaN      5039.0  1.180912e+21  1.094032e+18  0.000000e+00   
2          NaN      4662.0  3.349653e+17  8.868750e+15  6.942000e+18   
3          NaN      2575.0  9.319584e+17  4.381200e+16  0.000000e+00   
4          NaN      2428.0  3.942885e+18  3.579710e+17  3.716371e+20   
...        ...         ...           ...           ...           ...   
59894      NaN         1.0  4.497151e+16  0.000000e+00  0.000000e+00   
59895      NaN         1.0  8.237654e+14  0.000000e+00  0.000000e+00   
59896      NaN         1.0  8.253785e+16  0.000000e+00  0.000000e+00   
59897      NaN         NaN  0.000000e+00  0.000000e+00  0.000000e+00   
59898      NaN         NaN  4.006043e+18  0.000000e+00  0.000000e+00   

                 USDC           DAI           ASH           LUX    FIRST  \
0      10537174098434  5.102138e+23  1.107746e+20  6.812250e+21  7282250   
1           165402000  0.000000e+00  0.000000e+00  0.000000e+00        0   
2                   0  0.000000e+00  0.000000e+00  0.000000e+00        0   
3                   0  0.000000e+00  0.000000e+00  0.000000e+00        0   
4                   0  0.000000e+00  0.000000e+00  0.000000e+00        0   
...               ...           ...           ...           ...      ...   
59894               0  0.000000e+00  0.000000e+00  0.000000e+00        0   
59895               0  0.000000e+00  0.000000e+00  0.000000e+00        0   
59896               0  0.000000e+00  0.000000e+00  0.000000e+00        0   
59897               0  0.000000e+00  0.000000e+00  0.000000e+00        0   
59898               0  0.000000e+00  0.000000e+00  0.000000e+00        0   

          ARKE          ARTx          KLTR          TRSH          TATR  \
0      1500000  9.750650e+19  2.700000e+20  4.095000e+18  2.775177e+20   
1            0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
2            0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
3            0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
4            0  0.000000e+00  0.000000e+00  0.000000e+00  2.504000e+18   
...        ...           ...           ...           ...           ...   
59894        0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
59895        0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
59896        0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
59897        0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
59898        0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   

                ROT  ETH_base10  Occurrence_base2  
0      3.695128e+17   21.565755         17.591091  
1      0.000000e+00   21.072218         12.298922  
2      0.000000e+00   17.525000         12.186733  
3      0.000000e+00   17.969397         11.330357  
4      0.000000e+00   18.595814         11.245553  
...             ...         ...               ...  
59894  0.000000e+00   16.652937          0.000000  
59895  0.000000e+00   14.915804          0.000000  
59896  0.000000e+00   16.916653          0.000000  
59897  0.000000e+00        -inf               NaN  
59898  0.000000e+00   18.602716               NaN  

[59899 rows x 18 columns]

In [ ]:
df_binary = df_to_balance[["ETH_base10", 'Occurrence_base2']]
df_binary.head()

ETH_base10  Occurrence_base2
0   21.565755         17.591091
1   21.072218         12.298922
2   17.525000         12.186733
3   17.969397         11.330357
4   18.595814         11.245553

In [ ]:
df_binary.replace([np.inf, -np.inf], np.nan, inplace=True)
df_binary.dropna(subset=["ETH_base10", 'Occurrence_base2'], how="all", inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:5244: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,
/usr/local/lib/python3.7/dist-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [ ]:
sns.lmplot(x ="ETH_base10", y ="Occurrence_base2", data = df_binary, order = 2, ci = None)

NameError: ignored

In [ ]:
X = np.array(df_binary['ETH_base10']).reshape(-1, 1)
y = np.array(df_binary['Occurrence_base2']).reshape(-1, 1)
 
# Separating the data into independent and dependent variables
# Converting each dataframe into a numpy array
# since each dataframe contains only one column
 
# Dropping any rows with Nan values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)
 
# Splitting the data into training and testing data
regr = LinearRegression()
 
regr.fit(X_train, y_train)
print(regr.score(X_test, y_test))

**Since the linear correlation is low, we attempt the same analysis with smaller data set.**

**Top 500**

In [ ]:
df_binary500 = df_binary[:][:500]
   
# Selecting the 1st 500 rows of the data
sns.lmplot(x ="ETH_base10", y ="Occurrence_base2", data = df_binary500,
                               order = 2, ci = None)

In [ ]:
df_binary500.fillna(method ='ffill', inplace = True)
 
y = np.array(df_binary500['Occurrence_base2']).reshape(-1, 1)
X = np.array(df_binary500['ETH_base10']).reshape(-1, 1)
 
df_binary500.dropna(inplace = True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)
 
regr = LinearRegression()
regr.fit(X_train, y_train)
print(regr.score(X_test, y_test))

**Top 50**

In [ ]:
df_binary50 = df_binary[:][:50]
   
# Selecting the 1st 50 rows of the data
sns.lmplot(x ="ETH_base10", y ="Occurrence_base2", data = df_binary50,
                               order = 2, ci = None)

In [ ]:
df_binary50.fillna(method ='ffill', inplace = True)
 
y = np.array(df_binary50['Occurrence_base2']).reshape(-1, 1)
X = np.array(df_binary50['ETH_base10']).reshape(-1, 1)
 
df_binary50.dropna(inplace = True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)
 
regr = LinearRegression()
regr.fit(X_train, y_train)
print(regr.score(X_test, y_test))

In [ ]:
from sklearn.metrics import mean_absolute_error,mean_squared_error
 
mae = mean_absolute_error(y_true=y_test,y_pred=y_pred)
#squared True returns MSE value, False returns RMSE value.
mse = mean_squared_error(y_true=y_test,y_pred=y_pred) #default=True
rmse = mean_squared_error(y_true=y_test,y_pred=y_pred,squared=False)
 
print("MAE:",mae)
print("MSE:",mse)
print("RMSE:",rmse)